core.ts

```diff
@@ -476,9 +488,14 @@ const handleFilterOperator = (
 		case '$ceiling':
 		case '$isof':
 		case '$cast':
-			return filterFunction(filter, operator, parentKey);
+			return filterFunction(
+				filter as FilterType<typeof operator>,
+				operator,
+				parentKey,
+			);
 		// break
 		case '$raw': {
+			filter = filter as FilterType<typeof operator>;
```

Comment:
This is in order to narrow `filter` to the specific types it can be based on the operators we've matched, ideally I'd do it without any runtime code but I couldn't find a way, and my expectation is that `filter = filter` would be easily optimized away by any js engine